### PR TITLE
[Refactor] Convert Notebook to Scripts

### PR DESCRIPTION
**Pull request by agpeshal**
Clean-up old code with a better overall structure

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,66 +2,41 @@
...
-### Attack
 
+### Attack
+Download model [checkpoint](https://drive.google.com/file/d/1cMzKCL3Woa-oZzz6gQyP-pYYJ-L9tPPw/view?usp=sharing)
+To run PGD attack on model using CIFAR-10 images, run 
 ```bash
-cd Attack
-python cifar_model.py
-jupyter notebook
+cd source
```
Comment (abhiagg92) [inline README.md]:
```suggestion
cd src
```

File: README.md [hunk 2]
```diff
@@ -2,66 +2,41 @@
...
-python train.py --defense --num_epochs 20
-```
-
-This will train on MNIST and save the model the model as ```models/Net_{args.num_epochs}_{args.defense}``` 
-
-To evaluate the performance before and after adversarial training on attacked and non-attacked images, run the following command
-
 ```bash
-python eval.py --defense --num_epochs 20
+cd source
```
Comment (abhiagg92) [inline README.md]:
```suggestion
cd src
```

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+tk=8.6.12=h1ccaba5_0
+tomli=2.0.1=py310h06a4308_0
+torch=2.2.1+cpu=pypi_0
+torchvision=0.17.1+cpu=pypi_0
+tornado=6.1=py310h5764c6d_3
+tqdm=4.66.2=pypi_0
+traitlets=5.14.2=pyhd8ed1ab_0
+types-tqdm=4.66.0.20240106=pypi_0
+typing-extensions=4.9.0=py310h06a4308_1
+typing_extensions=4.9.0=py310h06a4308_1
```
Comment (abhiagg92) [inline requirements.txt]: These two are exactly same, maybe we can remove one from the file

File: src/attack.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import argparse
+
+import torch
+from torch.utils.data import DataLoader
+from torchvision import datasets, transforms
+
+from .attackers import PGD
```
Comment (abhiagg92) [inline src/attack.py]: I generally prefer absolute imports especially if import path is short. This will make it more readable
https://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module

File: src/attack.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            train=False,
+            download=True,
+            transform=transforms.ToTensor(),
+        ),
+        shuffle=True,
+        batch_size=1,
+    )
+
+
+def main() -> None:
```
Comment (abhiagg92) [inline src/attack.py:20]: Maybe we can `attack` as function name for better readability.

File: src/attack.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    args = parse_args()
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    model = CNNClassifier().to(device)
+
+    if args.ckpt is not None:
```
Comment (abhiagg92) [inline src/attack.py:26]: Will the attack work if checkpoint path is not given?

File: src/attack.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+from torch.utils.data import DataLoader
+from torchvision import datasets, transforms
+
+from .attackers import PGD
+from .net.cnn_classifier import CNNClassifier
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--ckpt", type=str, default=None)
```
Comment (abhiagg92) [inline src/attack.py:12]: Maybe add description of these arguments

File: src/adversarial_trainer.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        self.model.eval()
+        correct, total = 0, 0
+        for images, labels in tqdm(iter(self.test_loader), desc="Evaluating ..."):
+            images = images.to(self.device)
+            labels = labels.to(self.device)
+            predictions = self.predict_(images).detach()
+            correct += (labels == predictions).sum().cpu().numpy()
+            total += len(labels)
+
+        print("Test accuracy: {:.2f}".format(correct / total * 100.0))
```
Comment (abhiagg92) [inline src/adversarial_trainer.py]: Use f-formatted print statement wherever possible https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep498

File: src/adversarial_trainer.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    def predict_(self, images: torch.Tensor) -> torch.Tensor:
```
Comment (abhiagg92) [inline src/adversarial_trainer.py]: Use `_predict` instead of `predict_` as per the standard conventions to represent private method in python

File: src/adversarial_trainer.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        return torch.argmax(self.model(images).data, dim=1)
+
+    def train_epoch_(self) -> None:
```
Comment (abhiagg92) [inline src/adversarial_trainer.py]: same here

File: src/train.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    )
+    parser.add_argument("--steps", type=int, default=10, help="number of PGD steps")
+    parser.add_argument(
+        "--eval_interval", type=int, default=5, help="interval for inference"
+    )
+
+    return parser.parse_args()
+
+
+def main() -> None:
```
Comment (abhiagg92) [inline src/train.py:30]: You can also rename this method to be more descriptive

File: src/train.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    args = parse_args()
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    batch_size = args.bs
+    train_loader = DataLoader(
```
Comment (abhiagg92) [inline src/train.py]: create method to create both train and test dataloaders without code duplication

File: README.md (hunk 1 above)
Comment (agpeshal) [reply README.md]: fixed.

File: src/attack.py (hunk 5 above)
Comment (agpeshal) [reply src/attack.py:20]: I feel the name of the script suffices.

File: src/train.py (hunk 11 above)
Comment (agpeshal) [reply src/train.py:30]: same
